### DISCUSSION TITLE
help for type hints for mixin class

### DISCUSSION THREAD

**Discussion by prim** [microsoft/pyright#7923] [Q&A] [answered]
Hello.

I'm developing a game and its very heavy python code will have a mixin design similar to the one below.

I hope to mark the `self` parameter of `Bag.add_item` as Entity class. In this way, vscode can prompt to complete the `log_debug` function when I enter `self.log_`.

However, pyright will consider this kind of annotation illegal and report `Type of parameter "self" must be a supertype of its class "Log"`

Please help me, how should I write the type annotation for this mixin class? Or how to avoid pright reporting these errors.

```python

class Log(object):

	def log_debug(self: "Entity", msg):
		pass

class Bag(object):

	def add_item(self: "Entity"):
		self.log_debug("")

class Entity(Bag, Log):
	pass


```

**Comment by erictraut** [comment, marked as answer]
The way this is typically done is through the use of a protocol class.

```python
from typing import Protocol

class SupportsLogDebug(Protocol):
    def log_debug(self, msg: str): ...

class Log(SupportsLogDebug):
    def log_debug(self, msg: str):
        pass

class Bag:
    def add_item(self: SupportsLogDebug):
        self.log_debug("")

class Entity(Bag, Log):
    pass
```

Or if you don't want to define a separate protocol class, you could do:

```python
from typing import Protocol

class Log(Protocol):
    def log_debug(self, msg: str):
        pass

class Bag:
    def add_item(self: Log):
        self.log_debug("")

class Entity(Bag, Log):
    pass
```

In general, this isn't a pattern I would recommend in object-oriented programming. The way you've defined this class hierarchy, the parent class `Bag` is not usable on its own. It's useful only when combined with some other functionality. Refactoring the class definitions would avoid this.

**Comment by Azureblade3808** [comment]
You can have all your mix-in classes subclass `typing.Protocol` and leave other code unmodified.

```python
from typing import Protocol

class Log(Protocol):
    def log_debug(self: "Entity", msg):
        pass

class Bag(Protocol):
    def add_item(self: "Entity"):
        self.log_debug("")

class Entity(Bag, Log):
    pass
```
